pkg/codegen/schema/schema.go

```diff
@@ -1522,15 +1558,81 @@ type ResourceSpec struct {
 	Methods map[string]string `json:"methods,omitempty" yaml:"methods,omitempty"`
 }
 
+// ReturnTypeSpec is either ObjectTypeSpec or TypeSpec
+type ReturnTypeSpec struct {
+	ObjectTypeSpec *ObjectTypeSpec
+	TypeSpec       *TypeSpec
+}
+
+func (returnTypeSpec *ReturnTypeSpec) UnmarshalJSON(inputJSON []byte) error {
+	var data map[string]interface{}
+	if err := json.Unmarshal(inputJSON, &data); err != nil {
+		return err
+	}
+
+	if len(data) == 0 {
+		return nil
+	}
+
+	var objectSpec *ObjectTypeSpec
+	var typeSpec *TypeSpec
+	if _, hasProperties := data["properties"]; hasProperties {
+		if err := json.Unmarshal(inputJSON, &objectSpec); err != nil {
+			return err
+		}
+	} else {
+		if err := json.Unmarshal(inputJSON, &typeSpec); err != nil {
+			return err
+		}
+	}
+
+	returnTypeSpec.TypeSpec = typeSpec
+	returnTypeSpec.ObjectTypeSpec = objectSpec
+	return nil
+}
+
+func (returnTypeSpec *ReturnTypeSpec) UnmarshalYAML(node *yaml.Node) error {
```

Comment:
TIL this could take a `*yaml.Node`. 😄 

I think this function is identical to `UnmarshalJSON`. Could we set this up so we pass in a decoder function and do some type casting to avoid the duplication. Having two functions that should be the same module type seems error prone.

Reply:
> TIL this could take a *yaml.Node. 😄

Stole that from the implementation of `RawMessage` 😁 

So after a bit of testing, it looks like `UnmarshalYAML` is broken, i.e. it is not being called and doesn't work however I try. Changing the signature to `func (funcSpec *FunctionSpec) UnmarshalYAML(input []byte) error`yields the same results 😞

Reply:
Didn't need a to type case anything, a decoder function did the trick and added tests to go from and to JSON/YAML
```go
type Decoder func([]byte, interface{}) error

func (returnTypeSpec *ReturnTypeSpec) UnmarshalReturnTypeSpec(data []byte, decode Decoder) error {
   // implementation
}

func (returnTypeSpec *ReturnTypeSpec) UnmarshalJSON(inputJSON []byte) error {
	return returnTypeSpec.UnmarshalReturnTypeSpec(inputJSON, json.Unmarshal)
}

func (returnTypeSpec *ReturnTypeSpec) UnmarshalYAML(inputYAML []byte) error {
	return returnTypeSpec.UnmarshalReturnTypeSpec(inputYAML, yaml.Unmarshal)
}
```